### ISSUE TITLE
Rule Proposal: require-await

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#6820]
Async functions and await expressions arrived stage 4.
We can add the rule for those after our parser supported those.

This proposal is the `await` version of [require-yield](http://eslint.org/docs/rules/require-yield) rule.
It disallows async functions without await expressions.

``` js
/*eslint require-await: error*/

//------------------------------------------------------------------------------
// ✔ GOOD
async function foo() {
    await doSomething()
}

async function foo() {
    // do nothing.
}

//------------------------------------------------------------------------------
// ✘ BAD
async function foo() {
    doSomething()
}
```

**Comment by platinumazure** [comment]
Just for my curiosity, what is the impact of an asynchronous function with no await? Does it cause problems for the user or is it just a pointless keyword?

**Comment by not-an-aardvark** [comment]
It causes the function to return a Promise rather than a value.

```js
async function foo() {
  return 5;
}

// is equivalent to:

function foo() {
  return Promise.resolve(5);
}
```

**Comment by platinumazure** [comment]
:+1: from me, then.

**Comment by platinumazure** [comment]
@not-an-aardvark Wait, does that not also apply to functions with await as well? Or is it just that functions with await in them are more likely to be expected to be async/Promises and so this rule is just to try to help catch possible unintended use of async keyword?

**Comment by not-an-aardvark** [comment]
Yes, all async functions return Promises anyway, e.g.

```js
async function foo() {
  await fetchSomething();
  return 5;
}

// is equivalent to:

function foo() {
  return fetchSomething().then(function() {
    return 5;
  });
}
```

If the async function returns a Promise anyway, then the `async` keyword has no effect:

```js
function foo() {
  return Promise.resolve(5);
}

// is equivalent to:

async function foo() {
  return Promise.resolve(5); // this is a Promise anyway, so `async` makes no difference
}
```

One argument for *not* using this rule: `async` makes it clear that a function always returns a Promise, even if there are no await statements:

```js
async function foo() {
  // complicated logic without `await`
  return 5;
}

// is arguably clearer than:

function foo() {
  // complicated logic
  return Promise.resolve(5);
}
```

The first function in the last example is arguably clearer than the second function, because anyone who reads it will know that it always returns a Promise, without having to look for return values in the function body.

**Comment by platinumazure** [comment]
@not-an-aardvark Well, the awesome thing about ESLint is that users can choose not to use a rule if they don't want to. I can see this being useful for some folks.

**Comment by nzakas** [comment]
We don't yet support `async`, so it probably doesn't make sense to do this yet.

**Comment by michaelficarra** [comment]
I don't see what the actual benefit is here. Why would someone choose to enable this? It's not protecting them from any potential error, and it's not warning of something "unnecessary", as you cannot simply remove the `await` keyword and expect nothing to change.

**Comment by mysticatea** [comment]
@nzakas Yes. It's the reason I wrote, "We can add the rule for those after our parser supported those."

@michaelficarra Granted, `require-yield` rule does not say the reason. I think it's good to support refactoring. For an example, if I modify to a function which returns a `Promise` object from an async function, this rule will prevent some mistakes. Another example, if I just forgot to write `await` keyword, maybe this rule will tell it to me. I think an async function which not include `await` expressions has a smell.

**Comment by vutran** [comment]
I created a [plugin](https://github.com/vutran/eslint-plugin-async) for this a while back since I wasn't sure this could be packaged in here as it is using `babel-eslint` parser.

**Comment by nzakas** [comment]
I think this makes sense as a companion to `require-yield`, if you have an `async` function without `await`, you probably didn't mean to use `async`.

**Comment by platinumazure** [comment]
We have three :+1: (nzakas, hzoo, gyandeeps) and one :-1: (michaelficarra) from the team. If @michaelficarra can be persuaded that this rule is worth adding to core, we can accept the issue.

**Comment by nzakas** [comment]
@michaelficarra you're currently the only person dissenting. Are you willing to yield to the consensus here?

**Comment by ljharb** [comment]
generators and async functions aren't the same construct. I frequently use `async function` without `await`, because I don't like the serial style `await` forces, but I *do* like the two very strong guarantees that `async function` provides:
 1. it always returns a native `Promise`
 1. it never throws, ever, under any circumstances

Clearly I can simply choose not to enable this rule, but I think having the rule would actually cause a lot of harm by discouraging people from adding these guarantees to their promise-returning functions.

**Comment by nzakas** [comment]
@ljharb that's fair, there is a distinction that isn't immediately obvious.

>  I think having the rule would actually cause a lot of harm by discouraging people from adding these guarantees to their promise-returning functions.

I can't buy the argument that having a non-default rule available actively harms anyone. There are definitely arguments for not enabling the rule, but I don't see a good argument for not having the rule at all. 

I'd still bet that most people using `async` functions in most cases will do so with the intent of using `await`. For those who do not, then they just won't enable the rule.

**Comment by ljharb** [comment]
There's a number of rule proposals that have been excluded simply because of concerns about making core too large, and usually that's invoked when the rule isn't deemed compelling or beneficial enough.

I'm arguing here that people who enable the rule aren't doing themselves any favors, and that it provides zero actual benefit to anyone, because if you intend on using `await`, you'll use it, and if not, having an `async function` won't hurt in any way.

A generator without `yield` doesn't provide any benefit to anyone, and is almost certainly a bug - the same is not true for this case.

**Comment by not-an-aardvark** [comment]
I'm inclined to agree with @ljharb here; there are a number of valid cases for not using await in an async function:

```js
// this function is sync, but we want it to return a Promise for consistency
// or maybe it's conforming to an API that also allows async operations, so it must return a Promise
async function foo() {
  return fs.readFile
}
```

> I can't buy the argument that having a non-default rule available actively harms anyone. There are definitely arguments for not enabling the rule, but I don't see a good argument for not having the rule at all.

This would apply to pretty much any rule proposal. But if we're trying to keep core as small as possible and only allowing rules that ESLint is incomplete without, then I'm not convinced `require-await` meets the bar.

**Comment by nzakas** [comment]
@not-an-aardvark again, I don't think "there's a reason people won't use the rule" is a good argument against inclusion here. In that example, you could also just return `Promise.resolve('foo')` and get the same result. The consequence of enabling the rule is to say "are you sure you mean this function to be async?" This still seems, to me, like an area where people could get tripped up as it's not obvious that the intention of an async function that returns a value is intended to return a promise. We've always placed a higher value on helping people avoid mistakes and confusing code than we do on stylistic rules.

**Comment by not-an-aardvark** [comment]
My concern is that the rule seems to be advocating always using `await` in an async function as a best practice, even though there are a lot of cases where it's better to have an async function without using `await`.

In other words, while I agree that there are some cases<sup>1</sup> where this rule would be useful for preventing errors, there are a significant number of other cases<sup>2</sup>, even within the same codebase, where adherence to this rule would go against best practices and make code harder to read. This seems like a very undesirable quality for a linter rule. Unless the rule can distinguish between those cases automatically, I think enabling this rule on a project would be a footgun.

---

<sup>1</sup> Example of a case where the rule would be useful:

```js
async function foo() {
  const url = 'api.example.com/foo';
  const result = fetch(url); // oops, this was supposed to have `await`; the warning will be helpful
  return result.status === 200 ? 'request succeeded' : 'request failed';
}
```

<sup>2</sup> Example of a case where the rule would have a detrimental effect:

```js

// http middleware
// Middleware can be async, so all middleware functions must return a Promise.
module.exports = async function (req, res) {
  if (isValid(req.session)) {
    req.isAuthenticated = true;
  } else {
    throw new AuthenticationError();
  }
};
```

`require-await` would report the function above, and instead encourage writing it like this:

```js
// http middleware
// Middleware can be async, so all middleware functions must return a Promise.
module.exports = function (req, res) {
  if (isValid(req.session)) {
    req.isAuthenticated = true;
    return Promise.resolve();
  } else {
    return Promise.reject(new AuthenticationError());
  }
};
```

The second version makes it harder to verify that the function will always return a Promise. It also creates boilerplate with `Promise.resolve`/`Promise.reject`. Forgetting to wrap a value in `Promise.resolve` will cause a runtime error if the caller expects a Promise and calls `.then` on it. Also, if `isValid` throws, the middleware will throw a synchronous error rather than returning a rejected Promise.

**Comment by ljharb** [comment]
Using `await` in no way guarantees that users know that `async function` returns a promise - they may be under the misimpression that `await` is synchronous. Similarly, omitting `await` in no way demonstrates a lack of understanding that `async function` returns a promise.

**Comment by mysticatea** [comment]
@not-an-aardvark For example, if a library uses the following code to process middlewares, the middlewares don't need to return a promise because `await` expressions accept non-promise values. So we can remove the `async` keyword simply. 

```
async function compose(middlewares, args) {
    try {
        for (const middleware of middlewares) {
            await middleware(...args)
        }
    } catch (err) {
        processError(err)
    }
}
```

I think this is a companion of `require-yield`, and this is similar to `no-useless-bind` rule as well.
I think, if it must return a promise by some library-specific reason in rare cases, we can disable this rule.

**Comment by ljharb** [comment]
The difference is that a generator without `yield` has no use case, and a bind where it's useless has no use case - both rules simply prohibit almost-definitely-bugs.

An `async function` without `await` has a huge and common use case, and is almost never a bug.

In fact, in your example, if you remove `async` (thus forcing you to remove `await`), it wouldn't have the same behavior unless the loop was rewritten to be a `reduce` over `args` with a Promise accumulator.

**Comment by nzakas** [comment]
TSC Summary: The `require-await` rule enforces that async functions always contain `await`. It has a champion and three :+1:s as well as two :-1:s from the team. The argument against is that there are valid use cases for using async functions without `await`, so the rule might not be applicable in all situations. The argument for is that this rule is similar to `require-yield` for generator functions in that it seems more likely people will want to use `await` with async functions than not, and those who disagree can disable the rule.

TSC Question: Shall we accept this rule?

**Comment by ljharb** [comment]
@nzakas summary looks good with the addition that a generator without a yield has no use case, an async function without await has one.

**Comment by alberto** [comment]
There is also https://github.com/eslint/eslint/pull/7547 and there will be a conflict between the two, so if we are going to include this one, we should clarify how the two will interplay (ideally before merging either one).

**Comment by not-an-aardvark** [comment]
I don't see how these rules are in conflict with each other. Could you clarify what the conflict is?

**Comment by alberto** [comment]
If you enable both rules and you have:
```
async function foo() {
  return await bar;
}
```
`no-return-await` will error

If you change it to:
```
async function foo() {
  return bar;
}
```
`require-await` will error

**Comment by not-an-aardvark** [comment]
I see your point. However, I my understanding is that the `require-await` rule would instead encourage writing that function like this:

```js
function foo() {
  return Promise.resolve(bar);
}
```

So I agree that having both rules could lead to confusion, but I'm not sure they directly conflict. My opinion is that using `return await` is an antipattern regardless of whether `require-await` is used.

**Comment by ljharb** [comment]
Why would that be an improvement though? You shouldn't wrap your async function's return value in `Promise.resolve` for the same reason you shouldn't prepend `await` to it.

I'm standing by my claim that this rule should not exist, and is harmful if it does.

**Comment by not-an-aardvark** [comment]
I agree that wrapping it in `Promise.resolve` would not be an improvement. I was arguing that if this rule does get added, it shouldn't block `no-return-await`.

**Comment by ljharb** [comment]
There are a number of rules that if both turned on, conflict with each other - @alberto as long as these aren't both enabled in the recommended config, is there a reason that one of them need block the other?

**Comment by not-an-aardvark** [comment]
Yes, if they had a clear conflict with each other, we would probably want to avoid adding both of them to core. We generally try to avoid conflicts when we can.

**Comment by nzakas** [comment]
I don't think there's a clear conflict between these two rules. `require-await` can be resolved by adding an `await` at a spot other than right after `return`.

**Comment by alberto** [comment]
@nzakas I am not sure I understand. Can you show how you would fix the example I posted?

**Comment by ljharb** [comment]
@alberto 
```js
async function foo() {
  const baz = await bar;
  return baz;
}
```

That of course is contrived, just like the provided example, but if someone would choose to enable `require-await` they're not concerned with readability imo :-)

**Comment by btmills** [comment]
After discussion in the 2016-11-10 TSC meeting, the team decided to accept this rule into core.
